rfcs/inlang-git-server-tech-stack/RFC.md

```diff
@@ -0,0 +1,80 @@
+---
+# the frontmatter is only relevant for rendering this site on the website
+title: inlang git server tech stack
+description: The tech stack for the inlang git server
+---
+
+# inlang git server tech stack
+
+## Overview
+
+We will be building a (variant) of git-server soon as part of [the next git vision 2023](https://gist.github.com/araknast/d80f8b64b2777e42b49fd9e89e5a1b13).
+
+## Requirements
+
+- TypeScript to increase iteration speed (no Go, Rust, etc.) [High Confidence]
+
+Similar to our findings in [git-sdk-requirements](../git-sdk-requirements/RFC.md), we want to iterate quickly and TypeScript is the best way to do so.
+
+- No lock-in to a specific JS runtime (Node, Deno, Bun, Cloudflare workers, etc.) [High Confidence]
+
+Developers should have the possibility to choose their own JS runtime based on their requirements. This ensures that inlang-git caters to a wider range of developers and subsequently apps being build with it.
+
+- No lock-in to a specific deployment strategy (self-hosted, cloud, serverless etc.)
+
+Developers should have the possibility to choose their own deployment strategy based on their requirements. This ensures that inlang-git caters to a wider range of developers and subsequently apps being build with it.
+
+## Proposal
+
+All requirements seem to be fulfilled by having a "modern express" server that is runtime and deployment-independent. Deploying plain JS is extremly simple.
```

Comment:
You are right, we don't want to be locked into HTTP. 

My analogy of "a modern express" is referring to using cross-runtime JS web standards. Streaming over the network is finally supported (since a year), see https://developer.mozilla.org/en-US/docs/Web/API/Streams_API. If we choose a server framework like Hattip, we can extend the framework to support streams. Seems like a better path than writing our own server framework.

Reply:
In my experience, using high level HTTP server frameworks only makes sense for building complex HTTP server applications that serve web pages, and most of their functionality is catered to this. In my view, extending hattip to work with the `git://` protocol seems like ti would would be as much (or probably more) work than  building our on framework using raw sockets, based on something like https://github.com/websockets/ws.

Reply:
**Refocusing the discussion:** My main points in this thread are 

1. Using runtime agnostic JS to enable a variety of use cases.
2. Have a simple deployment model to ease adoption (if 1 is obeyed, 2 should be automatically obeyed as well). 

HTTP is the wrong anchor for the debate (although http3 seems to include a WebSocket replacement named WebTransport). Your argument that HTTP-focused servers cater to HTTP use cases is valid.

> our on framework using raw sockets, based on something like https://github.com/websockets/ws. 

Off-topic: Using this is a lock-in to Node.js and wouldn't obey 1 (if we deem a runtime agnostic stack to be of importance)

Reply:
> HTTP is the wrong anchor for the debate (although http3 seems to include a WebSocket replacement named WebTransport). Your argument that HTTP-focused servers cater to HTTP use cases is valid.

In that case I agree with the rest of your points. 

> Off-topic: Using this is a lock-in to Node.js and wouldn't obey 1 (if we deem a runtime agnostic stack to be of importance)

If we want a runtime agnostic framework using raw sockets, there is also https://github.com/heineiuo/isomorphic-ws.